src/vos/vos_internal.h

```diff
@@ -363,6 +369,31 @@ struct vos_container {
 	struct btr_root		vc_dtx_committed_btr;
 	/* The list for active DTXs, roughly ordered in time. */
 	d_list_t		vc_dtx_act_list;
+	/* The list for the active DTX entries with epoch sorted. */
+	d_list_t		vc_dtx_sorted_list;
+	/* The list for the active DTX entries (but not re-indexed) with epoch unsorted. */
+	d_list_t		vc_dtx_unsorted_list;
+	/* The list for the active DTX entries that are re-indexed when open the container. */
+	d_list_t		vc_dtx_reindex_list;
+	/* The largest epoch difference for re-indexed DTX entries max/min pairs. */
+	uint64_t		vc_dtx_reindex_eph_diff;
+	/* The latest calculated local stable epoch. */
+	daos_epoch_t		vc_local_stable_epoch;
+	/*
+	 * The lowest epoch boundary for current acceptable modification. It cannot be lower than
+	 * vc_local_stable_epoch, otherwise, it may break stable epoch semantics. Because current
+	 * target reported local stable epoch may be used as global stable epoch. There is window
+	 * between current target reporting the local stable epoch and related leader setting the
+	 * global stable epoch. If the modification with older epoch arrives during such internal,
+	 * we have to reject it to avoid potential conflict.
+	 *
+	 * On the other hand, it must be higher than EC/VOS aggregation up boundary. Under space
+	 * pressure, the EC/VOS aggregation up boundary may be higher than vc_local_stable_epoch,
+	 * then it will cause vc_mod_epoch_bound > vc_local_stable_epoch.
+	 */
+	daos_epoch_t		vc_mod_epoch_bound;
```

Comment:
I don't quite follow so many various variables, it looks to me there are only two things:
1. constant gap between current HLC and EC aggregation upper bound, any write with lower epoch (lower than aggregation upper bound) will be rejected. (write with "epoch < current HLC - gap" will be rejected)
2. per-container persistent stable epoch, which is min(uncommitted dtx epoch, agg upper bound epoch).

Did I miss anything?

Reply:
> 1. constant gap between current HLC and EC aggregation upper bound, any write with lower epoch (lower than aggregation upper bound) will be rejected. (write with "epoch < current HLC - gap" will be rejected)

Such way can work, but may cause too much transaction restart unnecessarily. In fact, we only need to guarantee that no modification will be older than aggregation boundary.

> 2. per-container persistent stable epoch, which is min(uncommitted dtx epoch, agg upper bound epoch).

That is only in theory, but it is not easy maintain DTX epoch order efficiently, we do not know which one is "min".

Reply:
I see why this variable was introduced now, though I don't think this optimization is necessary. :)